reconcile/ldap_groups/integration.py

```diff
@@ -0,0 +1,226 @@
+            raise ValueError("Duplicate ldapGroup value(s) found.")
+        return roles
+
+    def fetch_desired_state(
+        self, roles: Iterable[RoleV1], owners: Iterable[Entity], contact_list: str
+    ) -> list[Group]:
+        groups = []
+        for role in roles:
+            if not role.ldap_group:
+                # roles with empty ldap_group are already filtered out in get_roles; just make mypy happy
+                continue
+
+            groups.append(
+                Group(
+                    name=role.ldap_group,
+                    description="Persisted App-Interface role. Managed by qontract-reconcile",
+                    display_name=f"{role.ldap_group} (App-Interface))",
+                    members=[
+                        Entity(type=EntityType.USER, id=user.org_username)
+                        for user in role.users
+                    ],
+                    # only owners can modify the group (e.g. add/remove members)
+                    owners=owners,
+                    contact_list=contact_list,
+                )
+            )
+
+        return groups
+
+    def fetch_current_state(
+        self, internal_groups_client: InternalGroupsClient, group_names: Iterable[str]
+    ) -> list[Group]:
+        groups = []
+        for group_name in group_names:
+            try:
+                groups.append(internal_groups_client.group(group_name))
+            except NotFound:
+                pass
+        return groups
+
+    def reconcile(
+        self,
+        dry_run: bool,
+        internal_groups_client: InternalGroupsClient,
+        desired_groups: Iterable[Group],
+        current_groups: Iterable[Group],
+    ) -> None:
+        diff_result = diff_iterables(
+            current_groups,
+            desired_groups,
+            key=lambda g: g.name,
+            equal=lambda g1, g2: g1 == g2,
+        )
+        # Internal Groups API does not support listing all managed groups, therefore
+        # we need to keep track of them ourselves.
+        self._managed_groups = [g.name for g in current_groups]
+
+        for group_to_add in diff_result.add.values():
+            logging.info(
+                [
+                    "create_ldap_group",
+                    group_to_add.name,
+                    f"users={', '.join([u.id for u in group_to_add.members])}",
+                ]
+            )
+            if not dry_run:
+                internal_groups_client.create_group(group_to_add)
+                self._managed_groups.append(group_to_add.name)
+
+        for group_to_remove in diff_result.delete.values():
+            logging.info(["delete_ldap_group", group_to_remove.name])
+            if not dry_run:
+                try:
+                    internal_groups_client.delete_group(group_to_remove.name)
+                except NotFound:
+                    pass
+                self._managed_groups.remove(group_to_remove.name)
```

Comment:
I think this is negligible. It takes 0.1 secs on my Mac to remove an item from a list with 1000000 items.

Reply:
remove 1 is fast, if it's in a loop, it will be slow, try this:

```python
a = list(range(1000000))
for i in range(1000000):
  a.remove(i)
```

compare it with

```python
b = set(range(1000000))
for i in range(1000000):
  b.remove(i)
```

Reply:
👍 changed in b882a71e2c81ca44a1